File: 22-generate-parentheses/22-generate-parentheses.py
class Solution(object):
    def generateParenthesis(self, n):
        """
        :type n: int
        :rtype: List[str]
        """
        res = set()

        #n pairs of parentheses
        def combination(n, stack, temp):
            if n == 0 and not stack:
                res.add(temp)
                return
            #put '('
            if n > 0:
                stack.append('(')
                combination(n - 1, stack, temp + '(')
                stack.pop()
            #put ')'
            if stack:  
                popped = stack.pop()
                combination(n, stack, temp + ')')
                stack.append(popped)
        combination(n, [], '')
        return list(res)
            
            
                
            

File: 22-generate-parentheses/22-generate-parentheses.java
class Solution {

    public List<String> generateParenthesis(int n) {
        List<String> res = new LinkedList<>();
        this.generate(n, 0, new StringBuilder(), res);
        return res;
    }
    public void generate(int n, int rolling, StringBuilder temp, List<String> res){
        if (rolling == 0 && n == 0){
            res.add(temp.toString());
            return;
        }
        if(rolling < 0) return;
        if(n == 0){
            temp.append(')');
            this.generate(n, rolling-1, temp, res);
            temp.deleteCharAt(temp.length() - 1);
        }
        if(n > 0){
            temp.append('(');
            this.generate(n-1, rolling+1, temp, res);
            temp.deleteCharAt(temp.length() - 1);            
            temp.append(')');
            this.generate(n, rolling-1, temp, res);
            temp.deleteCharAt(temp.length() - 1);   
        }
        
        
        
    }
    
    
    
    
}